dvc/exceptions.py

```diff
@@ -251,16 +251,32 @@ def __init__(self, amount):
 
 
 class CheckoutError(DvcException):
-    def __init__(self, target_infos, stats=None):
+    def __init__(self, target_infos, stats=None, failed_not_stored=None):
         self.target_infos = target_infos
         self.stats = stats
-        targets = [str(t) for t in target_infos]
+        targets = []
+        for t in target_infos:
+            if failed_not_stored and t in failed_not_stored:
+                targets.append(
+                    "{} (marked with 'store: false' in dvc file)".format(t)
+                )
+            else:
+                targets.append(str(t))
+
         m = (
-            "Checkout failed for following targets:\n{}\nIs your "
+            "Checkout failed for following targets:\n{}\n\nIs your "
             "cache up to date?\n{}".format(
                 "\n".join(targets), error_link("missing-files"),
             )
         )
+
+        if failed_not_stored:
+            m = m + (
+                "\n\nNote that 'store: false' outputs are not "
+                "searched for in remote storage, only in the cache. "
+                "See {}".format(error_link("checkout-no-store"),)
```

Comment:
Wait... Why is this a "checkout" problem though? The checkout would fail with or without `store: false` simply because the files are not in the cache, no?

Fetching or pulling is where `store: false` matters as I understand this. So shouldn't it be a fetch error/hint (info) message?

Reply:
It's only the checkout that's a problem: `dvc fetch` happens fine because it doesn't attempt to retrieve the `store: false` files. It's only when you checkout (which is part of `dvc pull`) when it matters that they're not in the cache. 

So yes, it's the same as doing a checkout of a normal file that isn't in the cache. The additional error messages are just a hint to the user as to why that might have happened.

Reply:
> only the checkout that's a problem... The additional error messages are just a hint to the user as to why that might have happened.

Hmm.. If I just run checkout, I don't need to know anything about remote storage. These are basically unrelated things. The existing message about the files not being is the cache is enough there, I think. (Sorry, I know I already made you change it a lot 😅.)

> checkout (which is part of dvc pull

Good point. So pull itself is also not the issue.

I think this hint should happen independently at fetch, so that if you use fetch or pull and nothing is downloaded at least you get a message about a possible confusion. In the case of pull, if the files are also not in cache and you get that other (chekout) error, the combination of both messages should be what the user needs to know.

Reply:
But lets ask @efiop or/and @shcheklein to please confirm this before I make you go redo all the error handing again... Thanks!

Reply:
I don't think the hint can happen at fetch, because there's no error at that point: not pulling these files from the remote is the expected, intended behaviour so showing an error message would be strange. 

The error only occurs because the files aren't available locally when you checkout, so I think it's definitely a checkout error, not a fetch error. So I don't think there's anywhere you could show the hint other than at checkout, since that's the only place that an error happens.

Reply:
It's not an error, it would be an INFO message (a hint).

> not pulling these files from the remote is the expected
> I don't think there's anywhere you could show the hint other than at checkout

Shouldn't fetch be always checking the `store` value now anyway? So there case to print the hint should be there somewhere already.

I still think this situation has nothing to do with checkout, which should be completely disconnected from remote storage. @shcheklein @efiop can you please corroborate? Thanks

Reply:
This would be the place for a message on fetch:

https://github.com/iterative/dvc/blob/68471ebbf073b7c893ba3f9af17a09dc9425d761/dvc/output/base.py#L482-L490

Could be quite noisy though if a message here got fired for every import marked as `store: false`

Reply:
@pared what do you think about this discussion? Since I see you reviewed this from the eng. team. Thanks

Reply:
@jorgeorpinel 
As far as I understand from our most recent conversation (https://github.com/iterative/dvc/pull/4581#discussion_r509366337)

We are going to change the approach and hint user (via warning message) during `stage` checkout. Making this error message obsolete. @charlesbaynham am I right here?

Reply:
That's my understanding too. I just haven't done it yet!

Reply:
@charlesbaynham no worries, I am not trying to rush you, just clearing things up, so that we all are on the same page :)

Reply:
OK then, resolving here. Thanks